### DISCUSSION TITLE
Only UP/DOWN/ENTER button, how to edit values?

### DISCUSSION THREAD

**Discussion by SanderM2** [Spirik/GEM#69] [Q&A] [answered]
I'm trying to use GEM on an Industruino  (https://industruino.com) device that only has an UP, DOWN and ENTER button.
I'm succesfully running the Example-02_Blink.ino example on it right now but if I want to change the Interval, a variable. I can only change the first number of the value. So for example, change 500 to 400 or 300 etc...
But I cannot set it to 350 because I don't have a left or right button.

Is there a way make GEM cycle through the digits using only the UP/DOWN/ENTER buttons?
Like, when I press enter, it should cycle to the next digit, let me change it and continue until all 3 digits are changed?

Thanks

**Comment by Spirik** [comment]
Hello!

At the moment it may require some source code modification, I am afraid. For example, you can make `GEM._editValueMode` [property](https://github.com/Spirik/GEM/blob/master/src/GEM_u8g2.h#L182) public, and by checking it in your loop decide what to do when pressing Enter button: if `_editValueMode` is `true` then you are in edit mode and subsequent presses of Enter button (aka `GEM_KEY_ENTER`) should be interpreted as Right key  (aka `GEM_KEY_RIGHT`), unless the end of a variable string is reached (and it may be tricky to track that part).

Alternatively, you may want to consider avoiding editable variables altogether in favor of making use of Selects, Checkboxes and Buttons (if it is acceptable in your specific case, of course).

However I still plan on adding special means of tracking current state of GEM object (accessing currently selected Menu item object, its value, menu states, etc.) to make it a little bit easier to customize behavior of input methods. That will be added in some future versions of library, but no estimates as of now. You may see the discussion [here](https://github.com/Spirik/GEM/pull/42) for some additional context.

**Comment by Spirik** [comment]
Additional thing you can try to do (without the need to modify GEM source code) is to detect quick-time events (QTE) with what buttons you have available: e.g. quick double click of Enter/`GEM_KEY_ENTER` (or Up/`GEM_KEY_UP`, or Down/`GEM_KEY_DOWN`) button may be interpreted as `GEM_KEY_RIGHT` action.

For that you would have to write your own key detection method and call `menu.registerKeyPress` with action you would like to perform. 

Please, refer to [Control menu from your sketch](https://github.com/Spirik/GEM/wiki/How-to#control-menu-from-your-sketch) section of wiki for details.

**Comment by zandernc** [comment]
Hi - I'm rather new to all this so apologies in advance if any of this is clueless... but I'd like to add my voice in support of GEM working with just a rotary encoder and a single button - it's such an efficient UI input when you are trying to make a simple menu-based project. Maybe it's worth implementing? In my experience the difference between the "ok" button and the "right" button are pretty similar, and... I'm not sure you lose much functionality if you make them equivalent. Changing your source code is far beyond my skills... Thank you!

**Reply by Spirik** [reply]
Hi, **zandernc**!

It should be possible to use rotary encoder (or any other input device) with GEM without the need to modify source code (although edition of variables in this mode will require some creative solutions)! 

There is no example for rotary encoders specifically, but theoretically as soon as input device (or rather your sketch) can produce 6 basic operations that menu understands (up, right, down, left, ok, cancel) it should be completely possible. But it is up to you to figure out how to extract those readings from encoder (or any other input device for that matter). Presumably with the use of some 3rd party library, such as [Encoder](https://www.arduino.cc/reference/en/libraries/encoder/).

See [Control menu from your sketch](https://github.com/Spirik/GEM/wiki/How-to#control-menu-from-your-sketch) section of wiki for a little more details. You may want to look into how [`KeyDetector`](https://github.com/Spirik/GEM/wiki/How-to#detect-key-presses-with-keydetector) or [`U8g2`](https://github.com/Spirik/GEM/wiki/How-to#detect-key-presses-with-u8g2) key detection organized in supplied examples (in their `loop()`) as a starting point.

You also may be interested in the following discussions:
* #36 
* #61
* #57 (shows possible way to control GEM from custom device, in this case - Serial port)
* #71 (issue itself not related to the topic, but author provides listing of the code in which GEM is operated using rotary encoder)

Hope navigating through those resources helps!

**Reply by zandernc** [reply]
Just to follow up - I was able to implement a very functional version of the menu using an encoder with a button - turning the encoder for up and down keypresses, and a short press of the encoder for right keypress and a long press of the encoder for "ok." Works great! Thank you Spirik!

**Reply by Spirik** [reply]
One other possible solution, as I mention in #36, is to detect turning the encoder while pressed (may not be compatible with all of the encoders though). If it is supported by encoder, then I would suggest to consider the following scheme:
* Rotation - Up, Down;
* Rotation while pressed - Left, Right;
* Short press - Ok;
* Long press - Cancel.

That way we will cover all 6 interactions GEM is able to respond to.

**Reply by Spirik** [reply, marked as answer]
👋  FYI: just updated GEM with [examples](https://github.com/Spirik/GEM/wiki/Test-bench-with-rotary-encoder-for-U8g2-version) on how to control menu with rotary encoder using latest version of KeyDetector (implementing control scheme described above).
